source/server/server.cc

```diff
@@ -103,39 +103,11 @@ void InstanceImpl::failHealthcheck(bool fail) {
 }
 
 void InstanceUtil::flushMetricsToSinks(const std::list<Stats::SinkPtr>& sinks,
-                                       Stats::Store& store) {
+                                       Stats::FlushDelegate& flush_delegate) {
   for (const auto& sink : sinks) {
-    sink->beginFlush();
-  }
-
-  for (const Stats::CounterSharedPtr& counter : store.counters()) {
-    uint64_t delta = counter->latch();
-    if (counter->used()) {
-      for (const auto& sink : sinks) {
-        sink->flushCounter(*counter, delta);
-      }
-    }
-  }
-
-  for (const Stats::GaugeSharedPtr& gauge : store.gauges()) {
-    if (gauge->used()) {
-      for (const auto& sink : sinks) {
-        sink->flushGauge(*gauge, gauge->value());
-      }
-    }
-  }
-
-  for (const Stats::ParentHistogramSharedPtr& histogram : store.histograms()) {
-    if (histogram->used()) {
-      for (const auto& sink : sinks) {
-        sink->flushHistogram(*histogram);
-      }
-    }
-  }
-
-  for (const auto& sink : sinks) {
-    sink->endFlush();
+    sink->flush(flush_delegate);
   }
+  flush_delegate.clearFlushCache();
```

Comment:
Random thought maybe worthy of a TODO: I wonder if clearing the flush cache could somehow be done internal to the store only when new stats are added or something changes. At steady state, nothing will change, and we could avoid a bunch of recomputation on each flush interval. I think this is non-trivial due to locking required, but just throwing it out there.

Reply:
Another random thought: Does the flush cache add much value right now? I feel like most sinks are just going to access each vector once per interval? I wonder if we could remove the complexity for now or do you think it's worth it?

Reply:
On the first, I think that seems reasonable.  I'll add a TODO.

For the latter, I think it mainly adds value in the multi-sink case.  I was mainly trying to avoid perf regression for that use case.  I think the implementation of the caching is simple enough to be worth it since it can pretty much be entirely covered in a single test case, but I'm open to changing if that doesn't seem worthwhile.

Reply:
Caching for the multi-sink case makes sense. SGTM.